### PR TITLE
ENHANCEMENT Add version status badge in gridfield detail view

### PR DESCRIPTION
**Pull request by phalkunz**
Issue: https://github.com/silverstripe/silverstripe-versioned/issues/51

### CODE REVIEW

File: src/VersionedGridFieldItemRequest.php [hunk 1]
```diff
@@ -19,6 +19,23 @@
 class VersionedGridFieldItemRequest extends GridFieldDetailForm_ItemRequest
 {
 
+    public function Breadcrumbs($unlinked = false)
+    {
+        $items = parent::Breadcrumbs($unlinked);
+        if ($this->record->hasExtension(Versioned::class) && $this->record->config()->get('versioned_gridfield_extensions')) {
```
Comment (tractorcow) [inline src/VersionedGridFieldItemRequest.php]: I think we can remove the `versioned_gridfield_extensions` check; If you have this itemrequest class, you are already using the versioning extension for gridfield.

File: src/VersionedGridFieldItemRequest.php [hunk 2]
```diff
@@ -19,6 +19,23 @@
...
+            $status = $this->getRecordStatus();
+            $lastItem = $items->last();
+            if ($status) {
+                $badge = sprintf(
+                    '<span class="badge version-status version-status--%s">%s</span>',
```
Comment (tractorcow) [inline src/VersionedGridFieldItemRequest.php]: This is coming up as unreadable white text.

**Comment by tractorcow** [comment]
@phalkunz I've updated the code with the requested changes.

Also please see https://github.com/silverstripe/silverstripe-framework/pull/7773; In 4.1 I'm going to move this to versioned.

File: src/VersionedGridFieldItemRequest.php [hunk 3]
```diff
@@ -223,19 +226,32 @@ protected function setFormMessage($form, $message)
...
+     *
+     * @return array|null
+     */
     protected function getRecordStatus()
     {
+        /** @var DataObject|Versioned $record */
         $record = $this->record;
 
+        // No status if un-versioned
+        if (!$this->record->hasExtension(Versioned::class)) {
```
Comment (phalkunz) [inline src/VersionedGridFieldItemRequest.php]: This needs to check `versioned_gridfield_extensions` flag on the model because user can opt out.

Comment (tractorcow) [reply src/VersionedGridFieldItemRequest.php]: The opt out is done before VersionedGridFieldItemRequest is constructed.

```php
// Conditionally use a versioned item handler if it doesn't already have one.
        if ($record
            && $record->has_extension(Versioned::class)
            && $record->config()->get('versioned_gridfield_extensions')
            && (!$class || !is_subclass_of($class, VersionedGridFieldItemRequest::class))
        ) {
            $class = VersionedGridFieldItemRequest::class;
        }
```

Comment (tractorcow) [reply src/VersionedGridFieldItemRequest.php]: That was why my earlier request to move it here; We only had to check this once, and the condition would apply globally. :)

Comment (phalkunz) [reply src/VersionedGridFieldItemRequest.php]: Ah, right! 👍

File: src/VersionedGridFieldItemRequest.php [hunk 4]
```diff
@@ -19,6 +21,24 @@
...
+        if ($status) {
+            // Generate badge
+            $badge = DBField::create_field('HTMLFragment', sprintf(
+                '<span class="badge version-status version-status--%s">%s</span>',
+                $status['class'],
+                $status['title']
+            ));
+            /** @var ArrayData $lastItem */
+            $lastItem = $items->last();
+            $lastItem->setField('Extra', $badge);
```
Comment (flamerohr) [inline src/VersionedGridFieldItemRequest.php]: what happens if `Extra` was already defined by another class? Shouldn't this append rather than overwrite?

**Pull request merged**